feat: bulk book users to sessions via csv file

keevan opened the pull request:
Summary:
- permission check to only allow users with this permission to do so
- can intially set notification type, status (booked/cancelled) and
  discount code.
- tests for various scenarios.
- managed via a booking manager class.
- supports both csv file and array (list of bookings) formats.

Resolves #150

TODO: 
- [x] add event handling(?)

keevan committed:
feat: bulk book users to sessions via csv file

Summary:
- permission check to only allow users with this permission to do so
- can intially set notification type, status (booked/cancelled) and
  discount code.
- tests for various scenarios.
- managed via a booking manager class.
- supports both csv file and array (list of bookings) formats.

Resolves #150

keevan committed:
fix: phpcs issues and removed dead code

keevan commented:
Pulled in some changes found in 403, goal is that the behat issues have already been resolved.

Peterburnett commented on classes/booking_manager.php:
```diff
@@ -0,0 +1,316 @@
+            // Validate and get user.
+            $userids = $DB->get_records('user', ['email' => $entry->email], 'id');
+
+            // Multiple matched, ambiguous which is the real one.
```
Given that these have a couple of conditions on most of the fields, I reckon we should just split this up into a couple of private functions that just take a field -> errors, or row -> errors if there is cross dependencies. that way the main loop just reads as the series of checks being done at a high level

keevan replied:
I tested for user, and that was okay.., being that it worked when I added a method like `check_user($email, &$errors, $row) -> $userid`, but the problem is it introduced some level of inconsistency in how the checks are performed and recorded compared to sessions, which required different parameters to be provided, along with a session cache object, so it's probably better to keep all the checks inline for now, until there's a better path forward for all the checks collectively.

keevan replied:
For now I've left it inline and added a comment to the function (which was missing at the time)

Peterburnett commented on classes/form/confirm_bookings_form.php:
```diff
@@ -0,0 +1,67 @@
+
+        $backurl = new moodle_url('/mod/facetoface/upload.php', ['f' => $f]);
+        $htmlbuttons = $OUTPUT->render((new single_button(
+            new moodle_url('upload.php', ['f' => $f, 'fileid' => $fileid, 'process' => 1]),
```
one is an absolute URL, one is a relative, make these consistent?

Peterburnett commented on classes/form/upload_bookings_form.php:
```diff
@@ -0,0 +1,73 @@
+class upload_bookings_form extends \moodleform {
+
+    /**
+     * Build form for importing woekflows.
```
typo in comment

Peterburnett commented on classes/form/upload_bookings_form.php:
```diff
@@ -0,0 +1,73 @@
+
+        $mform->addElement('header', 'settingsheader', get_string('upload'));
+
+        $url = new moodle_url('example.csv');
```
Not sure if there is a convention or code style guide on this, but generally im partial to absolute URLs where we can, so the PHP file/function can be moved around without invisibly breaking a link to a static asset

keevan replied:
I blatantly copied the style from here:
https://github.com/moodle/moodle/blob/master/admin/tool/uploaduser/user_form.php#L43

But I can change it to absolute if preferred.

Peterburnett commented on upload.php:
```diff
@@ -0,0 +1,135 @@
+    throw new moodle_exception('error:incorrectcoursemoduleid', 'facetoface');
+}
+
+
```
double newline here

keevan replied:
Good catch

Peterburnett commented on upload.php:
```diff
@@ -0,0 +1,135 @@
+$modulecontext = context_module::instance($cm->id);
+require_capability('mod/facetoface:editsessions', $context);
+require_capability('mod/facetoface:uploadbookings', $context);
+
```
another one

Peterburnett commented on upload.php:
```diff
@@ -0,0 +1,135 @@
+        $bm->process();
+
+        echo "PROCESSED";
+        die;
```
leftover die?

keevan replied:
Whoops.

Peterburnett commented on upload.php:
```diff
@@ -0,0 +1,135 @@
+        // Redirect back to start with notification.
+        redirect(
+            new moodle_url('/mod/facetoface/upload.php', ['f' => $f]),
+            'Successfully processed X records.',
```
This bit looks like it wants to be a parameterised lang string?

keevan replied:
I made it a bit more general. The number of records processed would equal the number of rows in the file. I don't think that information is important as it's either all or nothing, as it wouldn't redirect if there was an exception during processing.

Peterburnett commented on upload.php:
```diff
@@ -0,0 +1,135 @@
+        notification::NOTIFY_ERROR);
+} else {
+    $mform = new upload_bookings_form(null);
+    $data = $mform->get_data();
```
is this get_data call actually needed?

keevan committed:
fix: clean up and some optimisations

- add missing phpdocs,
- optimise coursecontext get by caching the call to the class,
- convert moodle_url to abs path
- remove double blank lines
- remove dead code
- fix file upload handling

danmarsden commented:
I'm a bit confused here - is this a new feature? - or just a backport? 

new features should land in the latest branch first - then we can deal with the backports?

keevan commented:
> I'm a bit confused here - is this a new feature? - or just a backport?
> 
> new features should land in the latest branch first - then we can deal with the backports?

Good catch, I'll re-target the latest branch and create a backport PR once it's merged. Thanks Dan.

keevan committed:
style: phpcs fixes

Issues fixed:
      4 - Expected a space after cast statement

Peterburnett reviewed (approved):
I've largely reviewed the major commit, rather than the CI changes, but this one is looking good to go.

Peterburnett commented:
@danmarsden I've approved this one, its fairly self contained, did you have anything you wanted to mention before merge?

danmarsden commented:
if you've reviewed the full patch...and hard-coded english lang strings are fixed like this one I spotted:
https://github.com/catalyst/moodle-mod_facetoface/pull/151/files#diff-2359f56e8ab786049c0f9961df9bd8363b0f1e4e986972ebc692302ebc920000R119

then feel free to merge. :-)

Peterburnett commented on db/access.php:
```diff
@@ -30,160 +30,170 @@
+        ],
+    ],
+
+    // Ability to add a new facetoface activity to a course.
```
Comment needs updating

keevan replied:
Good spot. Fixed
